### PR TITLE
DFS implemented (using the BFS API)

### LINKED ISSUE #111: Add depth first search for graphs
#### Description of the problem

Serial and Parallel version of depth first search are required to be added for graphs.
#### References/Other comments
.. [1] https://en.wikipedia.org/wiki/Depth-first_search
.. [2] https://research.nvidia.com/sites/default/files/publications/nvr-2017-001.pdf

It looks like its too costly to implement parallel DFS as the algorithm is inherently sequential.

### PR DESCRIPTION
**Pull request by HarsheetKakar**
#### References to other Issues or PRs or Relevant literature

Fixes #111

### CODE REVIEW

File: pydatastructs/graphs/algorithms.py [hunk 1]
```diff
@@ -548,3 +551,84 @@ def strongly_connected_components(graph, algorithm):
         "isn't implemented for finding strongly connected components."
         %(algorithm, graph._impl))
     return getattr(algorithms, func)(graph)
+
+def depth_first_search(
+    graph, source_node, operation, *args, **kwargs):
+    """
+    Implementation of serial breadth first search(BFS)
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:558]:
```suggestion
    Implementation of serial depth first search (DFS)
```

File: pydatastructs/graphs/algorithms.py [hunk 2]
```diff
@@ -548,3 +551,84 @@ def strongly_connected_components(graph, algorithm):
...
+    algorithm.
+
+    Parameters
+    ==========
+
+    graph: Graph
+        The graph on which BFS is to be performed.
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:565]:
```suggestion
        The graph on which DFS is to be performed.
```

File: pydatastructs/graphs/algorithms.py [hunk 3]
```diff
@@ -548,3 +551,84 @@ def strongly_connected_components(graph, algorithm):
...
+    source_node: str
+        The name of the source node from where the BFS is
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:567]:
```suggestion
        The name of the source node from where the DFS is
```

File: pydatastructs/graphs/algorithms.py [hunk 4]
```diff
@@ -16,9 +16,12 @@
     'breadth_first_search_parallel',
     'minimum_spanning_tree',
     'minimum_spanning_tree_parallel',
-    'strongly_connected_components'
+    'strongly_connected_components',
+    'depth_first_search'
 ]
 
+Stack = Queue = deque
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:23]: :thumbsup:

File: pydatastructs/graphs/algorithms.py [hunk 5]
```diff
@@ -548,3 +551,84 @@ def strongly_connected_components(graph, algorithm):
...
+    dfs_stack = Stack()
+    visited = dict()
+    dfs_stack.append(source_node)
+    visited[source_node] = True
+    while len(dfs_stack) != 0:
+        curr_node = dfs_stack.pop()
+        next_nodes = graph.neighbors(curr_node)
+        if len(next_nodes) != 0:
+            for next_node in next_nodes:
+                if not next_node.name in visited:
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:623]:
```suggestion
                if visited.get(next_node.name, False) is False:
```

**Comment by HarsheetKakar** [comment]
sir is there any benefit of using .get method instead of simply checking the key

**Comment by czgdp1807** [comment]
In fact, `in` is better at least for numbers.
```python
>>> d = {k: v for k, v in zip(range(10**7), range(10**7))}
>>> import timeit
>>> def f():
...     return 10**6 in d
... 
>>> def g():
...     return d.get(10**6, None)
... 
>>> timeit.timeit(f)
0.09641253299923846
>>> timeit.timeit(g)
0.14119643199956045
```

File: pydatastructs/graphs/algorithms.py [hunk 6]
```diff
@@ -548,3 +551,84 @@ def strongly_connected_components(graph, algorithm):
...
+    dfs_stack = Stack()
+    visited = dict()
+    dfs_stack.append(source_node)
+    visited[source_node] = True
+    while len(dfs_stack) != 0:
+        curr_node = dfs_stack.pop()
+        next_nodes = graph.neighbors(curr_node)
+        if len(next_nodes) != 0:
+            for next_node in next_nodes:
+                if visited.get(next_node.name, False) is False:
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:623]:
```suggestion
                if next_node.name not in visited:
```

File: pydatastructs/graphs/algorithms.py [hunk 7]
```diff
@@ -548,3 +551,84 @@ def strongly_connected_components(graph, algorithm):
         "isn't implemented for finding strongly connected components."
         %(algorithm, graph._impl))
     return getattr(algorithms, func)(graph)
+
+def depth_first_search(
+    graph, source_node, operation, *args, **kwargs):
+    """
+    Implementation of serial depth first search (DFS)
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:558]:
```suggestion
    Implementation of depth first search (DFS)
```
Parallel version isn't possible for DFS.

**Comment by HarsheetKakar** [comment]
> In fact, `in` is better at least for numbers.
> 
> ```python
> >>> d = {k: v for k, v in zip(range(10**7), range(10**7))}
> >>> import timeit
> >>> def f():
> ...     return 10**6 in d
> ... 
> >>> def g():
> ...     return d.get(10**6, None)
> ... 
> >>> timeit.timeit(f)
> 0.09641253299923846
> >>> timeit.timeit(g)
> 0.14119643199956045
> ```

thats a good chunk of time saved, never thought it would have that.

File: pydatastructs/graphs/algorithms.py [hunk 8]
```diff
@@ -548,3 +551,84 @@ def strongly_connected_components(graph, algorithm):
...
+    >>> V2 = AdjacencyListGraphNode("V2")
+    >>> V3 = AdjacencyListGraphNode("V3")
+    >>> G = Graph(V1, V2, V3)
+    >>> from pydatastructs import depth_first_search
+    >>> def f(curr_node, next_node, dest_node):
+    ...     return curr_node != dest_node
+    ...
+    >>> G.add_edge(V1.name, V2.name)
+    >>> G.add_edge(V2.name, V3.name)
+    >>> depth_first_search(G, V1.name, f, V3.name)
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:601]:
```suggestion
    >>> depth_first_search(G, V1.name, f, V3.name)

    References
    =========

    .. [1] https://en.wikipedia.org/wiki/Depth-first_search
```

File: pydatastructs/graphs/algorithms.py [hunk 9]
```diff
@@ -548,3 +551,89 @@ def strongly_connected_components(graph, algorithm):
...
+
+    References
+    =========
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:604]:
```suggestion
    ==========
```

**Comment by czgdp1807** [comment]
Merging.

**Pull request merged**